core/domain/stats_jobs_one_off.py

```diff
@@ -18,14 +18,541 @@
+            # Compute num_completions for the states.
+            for session_id in (
+                    session_id_state_hit_event_mapping_for_this_version):
+                state_hit_event_dicts_sessioned = (
+                    session_id_state_hit_event_mapping_for_this_version[
+                        session_id])
+                # Sort list of state hit events in increasing order of
+                # timestamp.
+                state_hit_event_dicts_sessioned = sorted(
+                    state_hit_event_dicts_sessioned, key=lambda e: e[
+                        'created_on'])
+                # All state hit events in the above list are ordered by their
+                # timestamp. This implies that the last state hit event is the
+                # state at which the learner quit the exploration (This is only
+                # True if the last state hit event is not the terminal state).
+                # So, if the last state is not the terminal state, we increment
+                # num_completions for all states except the last state.
+                # Otherwise, we increment num_completions for all the states.
+                for state_hit_event in state_hit_event_dicts_sessioned[:-1]:
+                    state_stats_mapping[state_hit_event[
+                        'state_name']].num_completions_v1 += 1
+                if session_id in completed_session_ids:
+                    state_stats_mapping[state_hit_event_dicts_sessioned[-1][
+                        'state_name']].num_completions_v1 += 1
+
+            # Compute total_answers_count and useful_feedback_count.
+            for state_name in state_answer_counts_for_this_version:
+                state_stats_mapping[state_name].total_answers_count_v1 += (
+                    state_answer_counts_for_this_version[state_name][
+                        'total_answers_count'])
+                state_stats_mapping[state_name].useful_feedback_count_v1 += (
+                    state_answer_counts_for_this_version[state_name][
+                        'useful_feedback_count'])
+
+            # Now that the stats for the exploration's version are computed,
+            # calculate num_actual_starts. To compute this, we take the max of
+            # the first hit counts of all states that the initial state leads
+            # to. The max is taken instead of sum because the sum of
+            # first_hit_counts will sometimes end up counting a single learner
+            # twice. The max of both first hit counts ensures that we are
+            # counting only unique learners who traversed past the initial
+            # state.
+            init_state = versioned_exploration.states[
+                versioned_exploration.init_state_name]
+            max_first_hit_from_init_state = max([
+                state_stats_mapping[
+                    answer_group.outcome.dest].first_hit_count_v1
+                for answer_group in init_state.interaction.answer_groups if (
```

Comment:
Indent by 4. Move "if" part to next line, still indented by 4.

Reply:
That doesn't work? The 'for' needs to be aligned with state_stats_mapping (on the same indent level).

Reply:
Sorry! This one is my error. I misread.